src/stdfblib/ita/RMT_DEV.cpp

```diff
@@ -16,6 +17,15 @@
 #endif
 #include <stringdict.h>
 
+CDevice* CDevice::createDev(const std::string &paMGRID) {
```

Comment:
This seems like an inversion of responsibility when the base class directly creates an instance of a subclass. It also seems confusing to put a method belonging to `CDevice` in `RMT_DEV.cpp`.

How about having a separate `createDevice` in `CDevice` that chooses the correct implementation and delegates to a `createDevice` in the concrete class (e.g., `RMT_DEV`)? Maybe we could even use the typelib for this, since `CDevice` inherits from `CFunctionBlock`?

Reply:
The idea is to have a method that will create the respective device to be used for a specific 4diac FORTE instance. There will only be one sub-class of CDevice in any 4diac FORTE instance. Therefore I typelibrary will be an overkill and the wrong thing. As you see in may next commit this method will not be used by anyone else then CDevice and with that it is a kind of factory method that I can add to the code with compile time flags.

Reply:
I have created a PR based on this one with a slightly different solution that chooses the implementation based on a macro, see https://github.com/azoitl/4diac-forte/pull/1.

It avoids having to put `CDevice::createDev` into each of the concrete implementations and therefore allows to include all devices in compilation, meaning `OPCUA_DEV` can still inherit from `RMT_DEV`.

What do you think?

Reply:
I like the factory methods in the devices. It would open up in the future to select the device through a command line param. Although I don't know if we ever want to do that. 

What I don't find so nice is the macro based approach. I think a CMake code generation with `.in` files is better. What do you think @MartinMelikMerkumian?

`OPCUA_DEV` inheriting from RM_DEV was only an intermediate test for simplifying its development. One of the next steps would be to get rid of that (FYI @m-meingast).

Reply:
> Furthermore, it would open up in the future to select the device through a command line param. Although I don't know if we ever want to do that.

I believe that could make sense in situations when using a pre-compiled forte, such as when loading types with Lua or other means.

> What I don't find so nice is the macro based approach. I think a CMake code generation with .in files is better.

I changed the PR to use a configure file and type alias instead of the macro.

Reply:
I usually try to keep any configure file purely declarative if possible. I believe this makes them easier to read and maintain compared to having code mixed with replacements.

Reply:
I also think the whole `CDevice::createDev` should not be part of `CDevice`. 
I see at least 2 things trying to be solved which should not be part of `CDevice`, since it adds too much responsabiliy to it:
1. Creation of a device with `createDev`: maybe  you could have a `DeviceFactory` class/file handling the creation, and in it, you would have everything needed for handling (`.in`, dynamic or static polymorphism, ....). An idea could be to have it simple for now as:
  ```cpp
    #include "src/forte_device_config.h.in"

   std::unique_ptr<CDevice> createDevice()
   {
      return new FORTE_DEVICE(); 
   }

  ```
2. Handling the device life cycle: all these functions `startupNewDevice`,  `triggerDeviceShutdown` and `awaitDeviceShutdown` should be part of a `DeviceHandler` class with a private `std::unique_ptr<CDevice>` member which uses the factory from point 1. If at some point you want to instantiate based on command line paramaeters, the constructor of this handler should take this parameter, and the factory should be a dynamic factory based on this parameter.

Reply:
I like the IDE of a `C4diacFORTEInstance` class. Will, based on @mx990's improvement, rework it accordingly. At that point I find a DeviceFactory a bit an overkill and would suggest to keep it for now with the `C4diacFORTEInstance`.

Reply:
@cochicde I did a first prototype of your suggestion just for posix platform. If you think that is in the right direction I would implemented it across all architetures and clean-up the code.